File: lib/domain/weather_data/interactors/fetch_weather_data_perform_interactor.rb
# frozen_string_literal: true

module Domain
  module WeatherData
    module Interactors
      class FetchWeatherDataPerformInteractor
        include InputPorts::FetchWeatherDataPerformInputPort

        MAX_RETRY_ATTEMPTS = 5
        ALLOWED_MISSING_RATIO = 0.05
        SUFFICIENT_DATA_RATIO = 0.8

        def initialize(
          weather_data_gateway:,
          farm_gateway:,
          cultivation_plan_gateway:,
          agrr_weather_gateway:,
          presenter:
        )
          @weather_data_gateway = weather_data_gateway
          @farm_gateway = farm_gateway
          @cultivation_plan_gateway = cultivation_plan_gateway
          @agrr_weather_gateway = agrr_weather_gateway
          @presenter = presenter
        end

        def execute(input_dto:)
          latitude = input_dto[:latitude]
          longitude = input_dto[:longitude]
          start_date = input_dto[:start_date]
          end_date = input_dto[:end_date]
          farm_id = input_dto[:farm_id]
          cultivation_plan_id = input_dto[:cultivation_plan_id]
          channel_class = input_dto[:channel_class]
          executions = input_dto.fetch(:executions, 1)
          current_time = input_dto[:current_time]
          farm_info = farm_id ? "[Farm##{farm_id}]" : ""

          @presenter.info "🔍 [FetchWeatherDataJob] Received args: latitude=#{latitude}, longitude=#{longitude}, start_date=#{start_date}, end_date=#{end_date}, farm_id=#{farm_id}, cultivation_plan_id=#{cultivation_plan_id}, channel_class=#{channel_class}"

          # 日付検証
          if start_date.nil? || end_date.nil?
            error_msg = "Invalid date parameters: start_date=#{start_date.inspect}, end_date=#{end_date.inspect}"
            @presenter.error "❌ #{farm_info} #{error_msg}"
            raise ArgumentError, error_msg
          end

          period_str = start_date.year == end_date.year ? "#{start_date.year}" : "#{start_date.year}-#{end_date.year}"

          # フェーズ更新 (開始)
          if cultivation_plan_id && channel_class
            @cultivation_plan_gateway.update_phase(cultivation_plan_id, 'phase_fetching_weather', channel_class)
            @presenter.info "🌤️ [FetchWeatherDataJob] Started fetching weather data for plan ##{cultivation_plan_id}"
          end

          # 既存データチェック
          weather_location = @weather_data_gateway.find_weather_location_by_coordinates(latitude: latitude, longitude: longitude)
          if weather_location
            expected_days = (start_date..end_date).count
            existing_count = @weather_data_gateway.weather_data_count(
              weather_location_id: weather_location.id,
              start_date: start_date,
              end_date: end_date
            )
            threshold_days = (expected_days * SUFFICIENT_DATA_RATIO).ceil

            if existing_count >= threshold_days
              @presenter.info "⏭️  #{farm_info} Skipping #{period_str} - sufficient data exists (#{existing_count}/#{expected_days} days, #{((existing_count.to_f / expected_days) * 100).round(1)}%)"

              if farm_id
                @farm_gateway.increment_weather_data_progress(farm_id)
                progress = @farm_gateway.get_weather_data_progress(farm_id)
                fetched_years = @farm_gateway.get_weather_data_fetched_years(farm_id)
                total_years = @farm_gateway.get_weather_data_total_years(farm_id)
                @presenter.info "📊 #{farm_info} Progress: #{progress}% (#{fetched_years}/#{total_years} blocks)"
              end

              return
            end
          end

          @presenter.info "🌤️  #{farm_info} Fetching weather data for #{period_str} (#{latitude}, #{longitude})"

          # API負荷軽減のため短い待機時間を入れる
          sleep(0.5)

          # APIデータ取得
          weather_data = fetch_weather_from_agrr(latitude, longitude, start_date, end_date, farm_id)

          unless weather_data.is_a?(Hash)
            raise StandardError, 'Weather data response is invalid or missing'
          end

          data_points = weather_data['data']
          unless data_points.is_a?(Array)
            raise StandardError, 'Weather data response is invalid or missing'
          end

          expected_days = (start_date..end_date).count
          actual_days = data_points.size
          missing_days = [expected_days - actual_days, 0].max
          allowed_missing_days = (expected_days * ALLOWED_MISSING_RATIO).ceil

          if data_points.empty?
            raise StandardError, "Weather data missing for #{period_str} (0/#{expected_days} days)"
          elsif missing_days > allowed_missing_days
            raise StandardError, "Weather data missing #{missing_days} days exceeds allowed #{allowed_missing_days} days (#{(ALLOWED_MISSING_RATIO * 100).round(1)}%)"
          elsif missing_days.positive?
            @presenter.warn "⚠️  #{farm_info} Weather data incomplete for #{period_str}: #{actual_days}/#{expected_days} days (missing #{missing_days}, allowed #{allowed_missing_days})"
          end

          # WeatherLocation作成/取得
          location_data = weather_data['location']
          unless location_data.is_a?(Hash)
            raise StandardError, 'Weather data is missing location information'
          end
          weather_location = @weather_data_gateway.find_or_create_weather_location(
            latitude: location_data['latitude'],
            longitude: location_data['longitude'],
            elevation: location_data['elevation'],
            timezone: location_data['timezone']
          )

          # Farmリンク
          if farm_id
            @farm_gateway.update_weather_location_id(farm_id, weather_location.id)
            @presenter.info "🔗 [Farm##{farm_id}] Linked to WeatherLocation##{weather_location.id}"
          end

          # DTO作成 & upsert
          all_records = []
          data_points.each_with_index do |daily_data, index|
            date = Date.parse(daily_data['time'])
            record_attrs = {
              weather_location_id: weather_location.id,
              date: date,
              temperature_max: daily_data['temperature_2m_max'],
              temperature_min: daily_data['temperature_2m_min'],
              temperature_mean: daily_data['temperature_2m_mean'],
              precipitation: daily_data['precipitation_sum'],
              sunshine_hours: daily_data['sunshine_hours'],
              wind_speed: daily_data['wind_speed_10m'],
              weather_code: daily_data['weather_code'],
              updated_at: current_time
            }
            all_records << record_attrs

            if index == 0 || index == data_points.length - 1
              @presenter.debug "💾 [Weather Data ##{index + 1}] date=#{date}, temp=#{record_attrs[:temperature_min]}~#{record_attrs[:temperature_max]}°C"
            end
          end

          if all_records.any?
            dtos = all_records.map { |attrs| Domain::WeatherData::Dtos::WeatherDataDto.from_attrs(attrs) }
            @weather_data_gateway.upsert_weather_data!(
              weather_data_dtos: dtos,
              weather_location_id: weather_location.id
            )
          end

          data_count = all_records.size
          @presenter.info "💾 [Weather Data Summary] Total: #{data_count} records upserted in single batch"

          # Farm進捗更新
          if farm_id
            @farm_gateway.increment_weather_data_progress(farm_id)
            progress = @farm_gateway.get_weather_data_progress(farm_id)
            fetched_years = @farm_gateway.get_weather_data_fetched_years(farm_id)
            total_years = @farm_gateway.get_weather_data_total_years(farm_id)
            @presenter.info "📊 #{farm_info} Progress: #{progress}% (#{fetched_years}/#{total_years} blocks)"
          end

          @presenter.info "✅ #{farm_info} Saved #{data_count} weather records for #{period_str}"

          # フェーズ更新 (完了)
          if cultivation_plan_id && channel_class
            @cultivation_plan_gateway.update_phase(cultivation_plan_id, 'phase_weather_data_fetched', channel_class)
            @presenter.info "🌤️ [FetchWeatherDataJob] Weather data fetching completed for plan ##{cultivation_plan_id}"
          end
        end

        private

        def fetch_weather_from_agrr(latitude, longitude, start_date, end_date, farm_id)
          data_source = determine_data_source(farm_id, latitude: latitude, longitude: longitude)
          farm_info = farm_id ? "[Farm##{farm_id}]" : ""
          @presenter.info "🌍 #{farm_info} Using data source: #{data_source}"

          @agrr_weather_gateway.fetch_by_date_range(
            latitude: latitude,
            longitude: longitude,
            start_date: start_date,
            end_date: end_date,
            data_source: data_source
          )
        end

        def determine_data_source(farm_id, latitude:, longitude:)
          farm_entity = farm_id && @farm_gateway.find_by_id(farm_id) rescue nil

          if farm_entity
            return 'jma' if farm_entity.region == 'jp'
            lat = latitude
            lon = longitude
            return 'jma' if japan_location?(lat, lon)
            return 'nasa-power' if farm_entity.region.nil?
            return 'noaa'
          end

          lat = latitude
          lon = longitude
          return 'jma' if japan_location?(lat, lon)
          'noaa'
        end

        def japan_location?(latitude, longitude)
          return false if latitude.nil? || longitude.nil?
          latitude.between?(24.0, 46.0) && longitude.between?(130.0, 146.0)
        end

      end
    end
  end
end

File: crates/agrr-domain/src/weather_data/interactors/fetch_weather_data_perform_interactor.rs
//! Ruby: `Domain::WeatherData::Interactors::FetchWeatherDataPerformInteractor`

use serde_json::Value;
use time::Date;

use crate::shared::exceptions::RecordNotFoundError;
use crate::weather_data::dtos::{FetchWeatherDataPerformInput, WeatherData};
use crate::weather_data::gateways::{
    AgrrWeatherGateway, WeatherDataFarmGateway, WeatherDataGateway,
};
use crate::weather_data::helpers::parse_iso_date;
use crate::weather_data::ports::{
    FetchWeatherAdvancePhasePort, FetchWeatherDataJobPresenterPort, FetchWeatherPhase,
    RecordFarmWeatherBlockCompletedPort,
};

const SUFFICIENT_DATA_RATIO: f64 = 0.8;
const ALLOWED_MISSING_RATIO: f64 = 0.05;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FetchWeatherDataPerformError {
    InvalidWeatherApiResponse,
    InvalidWeatherDataArray,
    ExcessiveMissingWeatherDays,
    MissingOrInvalidWeatherLocation,
    InvalidDateParameters,
    WeatherDataStorageFailed(String),
}

impl std::fmt::Display for FetchWeatherDataPerformError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::WeatherDataStorageFailed(msg) => write!(f, "{msg}"),
            other => write!(f, "{other:?}"),
        }
    }
}

impl std::error::Error for FetchWeatherDataPerformError {}

/// Ruby: `Domain::WeatherData::Interactors::FetchWeatherDataPerformInteractor`
pub struct FetchWeatherDataPerformInteractor<'a> {
    weather_data_gateway: &'a dyn WeatherDataGateway,
    farm_gateway: &'a dyn WeatherDataFarmGateway,
    advance_phase: &'a dyn FetchWeatherAdvancePhasePort,
    record_block_completed: &'a dyn RecordFarmWeatherBlockCompletedPort,
    agrr_weather_gateway: &'a dyn AgrrWeatherGateway,
    presenter: &'a dyn FetchWeatherDataJobPresenterPort,
    skip_api_sleep: bool,
}

impl<'a> FetchWeatherDataPerformInteractor<'a> {
    pub fn new(
        weather_data_gateway: &'a dyn WeatherDataGateway,
        farm_gateway: &'a dyn WeatherDataFarmGateway,
        advance_phase: &'a dyn FetchWeatherAdvancePhasePort,
        record_block_completed: &'a dyn RecordFarmWeatherBlockCompletedPort,
        agrr_weather_gateway: &'a dyn AgrrWeatherGateway,
        presenter: &'a dyn FetchWeatherDataJobPresenterPort,
    ) -> Self {
        Self {
            weather_data_gateway,
            farm_gateway,
            advance_phase,
            record_block_completed,
            agrr_weather_gateway,
            presenter,
            skip_api_sleep: false,
        }
    }

    #[cfg(test)]
    pub fn with_skip_api_sleep(mut self) -> Self {
        self.skip_api_sleep = true;
        self
    }

    pub fn call(&self, input: FetchWeatherDataPerformInput) -> Result<(), FetchWeatherDataPerformError> {
        if input.start_date > input.end_date {
            return Ok(());
        }

        if input.cultivation_plan_id.is_some() && input.channel_class.is_some() {
            self.advance_phase.call(
                input.cultivation_plan_id.expect("checked"),
                FetchWeatherPhase::FetchingWeather,
                input.channel_class.as_deref().expect("checked"),
            );
        }

        if let Some(location) = self
            .weather_data_gateway
            .find_by_coordinates(input.latitude, input.longitude)
        {
            let expected_days = (input.end_date - input.start_date).whole_days() + 1;
            let existing_count = self
                .weather_data_gateway
                .weather_data_count(
                    location.id,
                    Some(input.start_date),
                    Some(input.end_date),
                )
                .map_err(|e| {
                    FetchWeatherDataPerformError::WeatherDataStorageFailed(e.to_string())
                })?;
            let threshold_days = (expected_days as f64 * SUFFICIENT_DATA_RATIO).ceil() as i64;

            if existing_count >= threshold_days {
                if let Some(farm_id) = input.farm_id {
                    let _ = self.record_block_completed.call(farm_id, input.current_time);
                }
                return Ok(());
            }
        }

        if !self.skip_api_sleep {
            std::thread::sleep(std::time::Duration::from_millis(500));
        }

        let weather_data = self.fetch_weather_from_agrr(
            input.latitude,
            input.longitude,
            input.start_date,
            input.end_date,
            input.farm_id,
        )?;

        let Some(weather_data) = weather_data else {
            return self.complete_fetch_without_new_data(&input);
        };

        let data_points = weather_data
            .get("data")
            .and_then(|v| v.as_array())
            .ok_or(FetchWeatherDataPerformError::InvalidWeatherDataArray)?;

        let expected_days = (input.end_date - input.start_date).whole_days() + 1;
        let actual_days = data_points.len() as i64;
        let missing_days = (expected_days - actual_days).max(0);
        let allowed_missing_days = (expected_days as f64 * ALLOWED_MISSING_RATIO).ceil() as i64;

        if missing_days > allowed_missing_days {
            return Err(FetchWeatherDataPerformError::ExcessiveMissingWeatherDays);
        } else if missing_days > 0 {
            self.presenter.warn("weather data incomplete");
        }

        let location_data = weather_data
            .get("location")
            .and_then(|v| v.as_object())
            .ok_or(FetchWeatherDataPerformError::MissingOrInvalidWeatherLocation)?;

        let weather_location = self
            .weather_data_gateway
            .find_or_create_weather_location(
                location_data
                    .get("latitude")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(input.latitude),
                location_data
                    .get("longitude")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(input.longitude),
                location_data.get("elevation").and_then(|v| v.as_f64()),
                location_data
                    .get("timezone")
                    .and_then(|v| v.as_str()),
            )
            .map_err(|_| FetchWeatherDataPerformError::MissingOrInvalidWeatherLocation)?;

        if let Some(farm_id) = input.farm_id {
            let _ = self
                .farm_gateway
                .update_weather_location_id(farm_id, weather_location.id);
        }

        let dtos: Vec<WeatherData> = data_points
            .iter()
            .filter_map(|daily| parse_weather_dto(daily))
            .collect();

        if !dtos.is_empty() {
            self.weather_data_gateway
                .upsert_weather_data(&dtos, weather_location.id)
                .map_err(|e| {
                    FetchWeatherDataPerformError::WeatherDataStorageFailed(e.to_string())
                })?;
        }

        if let Some(farm_id) = input.farm_id {
            let _ = self.record_block_completed.call(farm_id, input.current_time);
        }

        if input.cultivation_plan_id.is_some() && input.channel_class.is_some() {
            self.advance_phase.call(
                input.cultivation_plan_id.expect("checked"),
                FetchWeatherPhase::WeatherDataFetched,
                input.channel_class.as_deref().expect("checked"),
            );
        }

        Ok(())
    }

    pub fn determine_data_source(
        &self,
        farm_id: Option<i64>,
        latitude: f64,
        longitude: f64,
    ) -> String {
        let farm_entity = farm_id.and_then(|id| match self.farm_gateway.find_by_id(id) {
            Ok(entity) => Some(entity),
            Err(RecordNotFoundError) => None,
        });

        if let Some(farm) = farm_entity {
            if farm.region.as_deref() == Some("jp") {
                return "jma".to_string();
            }
            if farm.region.as_deref() == Some("in") {
                return "nasa-power".to_string();
            }
            if japan_location(latitude, longitude) {
                return "jma".to_string();
            }
            if farm.region.is_none() {
                return "nasa-power".to_string();
            }
            return "noaa".to_string();
        }

        if japan_location(latitude, longitude) {
            "jma".to_string()
        } else {
            "noaa".to_string()
        }
    }

    fn fetch_weather_from_agrr(
        &self,
        latitude: f64,
        longitude: f64,
        start_date: Date,
        end_date: Date,
        farm_id: Option<i64>,
    ) -> Result<Option<Value>, FetchWeatherDataPerformError> {
        let data_source = self.determine_data_source(farm_id, latitude, longitude);
        self.agrr_weather_gateway
            .fetch_by_date_range(latitude, longitude, start_date, end_date, &data_source)
            .map_err(|_| FetchWeatherDataPerformError::InvalidWeatherApiResponse)
    }

    /// agrr normal fetch exit 0 with no output file: keep existing store and continue the chain.
    fn complete_fetch_without_new_data(
        &self,
        input: &FetchWeatherDataPerformInput,
    ) -> Result<(), FetchWeatherDataPerformError> {
        if let Some(farm_id) = input.farm_id {
            let _ = self.record_block_completed.call(farm_id, input.current_time);
        }

        if input.cultivation_plan_id.is_some() && input.channel_class.is_some() {
            self.advance_phase.call(
                input.cultivation_plan_id.expect("checked"),
                FetchWeatherPhase::WeatherDataFetched,
                input.channel_class.as_deref().expect("checked"),
            );
        }

        Ok(())
    }
}

fn japan_location(latitude: f64, longitude: f64) -> bool {
    (24.0..=46.0).contains(&latitude) && (130.0..=146.0).contains(&longitude)
}

fn parse_weather_dto(daily: &Value) -> Option<WeatherData> {
    let date_str = daily.get("time")?.as_str()?;
    let date = parse_iso_date(date_str)?;
    Some(WeatherData::new(
        date,
        daily.get("temperature_2m_max").and_then(|v| v.as_f64()),
        daily.get("temperature_2m_min").and_then(|v| v.as_f64()),
        daily.get("temperature_2m_mean").and_then(|v| v.as_f64()),
        daily.get("precipitation_sum").and_then(|v| v.as_f64()),
        daily.get("sunshine_hours").and_then(|v| v.as_f64()),
        daily.get("wind_speed_10m").and_then(|v| v.as_f64()),
        daily
            .get("weather_code")
            .and_then(|v| v.as_i64())
            .map(|v| v as i32),
    ))
}

#[cfg(test)]
mod interactors_fetch_weather_data_perform_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/weather_data/interactors_fetch_weather_data_perform_interactor_test.rs"));
}
